examples/cookbook/langgraph_vs_mirascope/quickstart.py

```diff
@@ -0,0 +1,134 @@
+"""Mirascope implementation of LangGraph quickstart.
+
+Here is the reference link to LangGraph quickstart:
+    https://langchain-ai.github.io/langgraph/tutorials/introduction
+"""
+
+import requests
+from bs4 import BeautifulSoup
+from duckduckgo_search import DDGS
+from openai.types.chat import ChatCompletionMessageParam
+from pydantic import BaseModel, Field
+
+from mirascope.core import openai, prompt_template
+
+
+class WebSearch(openai.OpenAITool):
+    """Search the web for the given text and parse the paragraphs of the results."""
+
+    query: str = Field(..., description="The text to search for.")
+
+    def call(self) -> str:
+        """Search the web for the given text and parse the paragraphs of the results.
+
+        Returns:
+            Parsed paragraphs of each of the webpages, separated by newlines.
+        """
+        try:
+            # Search the web for the given text
+            results = DDGS(proxy=None).text(self.query, max_results=2)
+
+            # Parse the paragraphs of each resulting webpage
+            parsed_results = []
+            for result in results:
+                link = result["href"]
+                try:
+                    response = requests.get(link)
+                    soup = BeautifulSoup(response.content, "html.parser")
+                    parsed_results.append(
+                        "\n".join([p.text for p in soup.find_all("p")])
+                    )
+                except Exception as e:
+                    parsed_results.append(
+                        f"{type(e)}: Failed to parse content from URL {link}"
+                    )
+
+            return "\n\n".join(parsed_results)
+
+        except Exception as e:
+            return f"{type(e)}: Failed to search the web for text"
+
+
+class RequestAssistance(openai.OpenAITool):
+    """A tool that requests assistance from a human expert."""
+
+    query: str = Field(..., description="The users question.")
+
+    def call(self) -> str:
+        """Prompts a human to enter a response."""
+        print(f"The user has requested assistance. Here is the question: {self.query}")
```

Comment:
nit: do you mean the assistant has requested assistance?

Reply:
oh oops, yeah